server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -569,21 +604,69 @@ private void createShard(DiscoveryNodes nodes, RoutingTable routingTable, ShardR
                 failedShardHandler,
                 this::updateGlobalCheckpointForShard,
                 retentionLeaseSyncer,
-                nodes.getLocalNode(),
+                originalState.nodes().getLocalNode(),
                 sourceNode
             );
-        } catch (Exception e) {
-            failAndRemoveShard(shardRouting, true, "failed to create shard", e, state);
+            listener.onResponse(true);
+        } catch (ShardLockObtainFailedException e) {
+            if (e.getCause() instanceof InterruptedException || Thread.currentThread().isInterrupted()) {
+                logger.warn(Strings.format("interrupted while creating shard [{}]", shardRouting), e);
+                listener.onFailure(e);
+                return;
+            }
+            logger.log(
+                (iteration + 25) % 30 == 0 ? Level.WARN : Level.DEBUG,
+                "shard lock currently unavailable for [{}], retrying in [{}]: [{}]",
+                shardRouting,
+                shardLockRetryInterval,
+                e.getMessage()
+            );
+            // TODO could we instead subscribe to the shard lock and trigger the retry exactly when it is released rather than polling?
+            threadPool.scheduleUnlessShuttingDown(
+                shardLockRetryInterval,
```

Comment:
You may have already thought about this, but do you think we should introduce some jitter here? In case many shards were created after the same cluster state update and failed to immediately obtain their locks, they will al be retried at around the same time. (I'm mostly thinking about primary frozen shards after a node failure)

Reply:
That's ok I think, the number of concurrently `INITIALIZING` shards on this node is bounded by the `ThrottlingAllocationDecider`, and we're only scheduling tasks that add the retries to the applier thread's queue. The retries themselves execute one-at-a-time.